Scroll To Text Fragement re-loads after every dynamic style sheet load, and fights user scrolls to do so.

megangardner opened the pull request:
#### 2dd5d4cec216fe15f1571fda3d2b578056e5ce75
<pre>
Scroll To Text Fragement re-loads after every dynamic style sheet load, and fights user scrolls to do so.
<a href="https://bugs.webkit.org/show_bug.cgi?id=264195">https://bugs.webkit.org/show_bug.cgi?id=264195</a>
<a href="https://rdar.apple.com/112608578">rdar://112608578</a>

Reviewed by Simon Fraser.

In order to account for style changes adjusting the scroll of the page, after each dynamic style
sheet load, we recall scrollToFragment to set the page at the correct position again.

We need to stop automatically scrolling after the user has scrolled, as user scroll should trump
all URL programatic scrolls. So we reset the flag that causes scrolls to happen when a user scroll is
detected.

* LayoutTests/http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load-expected.txt: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async initiateUserScroll):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didRemoveAllPendingStylesheet):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):
(WebCore::LocalFrameView::setWasScrolledByUser):
* Source/WebCore/page/LocalFrameView.h:

Canonical link: <a href="https://commits.webkit.org/272151@main">https://commits.webkit.org/272151@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e9957e1a66c7d746c07a3f862023215d8a253bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9416 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33232 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27781 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31459 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11719 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6671 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31050 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27506 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6776 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27382 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34569 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27979 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27864 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33088 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6966 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5058 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30920 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8683 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27164 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7268 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7684 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7521 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -89,6 +89,7 @@ class View;
 Pagination::Mode paginationModeForRenderStyle(const RenderStyle&);
 
 enum class LayoutViewportConstraint : bool { Unconstrained, ConstrainedToDocumentRect };
+enum class ScrollToTextFragmentIndicator : bool { Disabled, Enabled };
```
I think this should be more like `ShowScrollToTextFragmentIndicator { No, Yes }`

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -4772,6 +4772,8 @@ void LocalFrameView::setWasScrolledByUser(bool wasScrolledByUser)
     cancelScheduledScrolls();
     if (currentScrollType() == ScrollType::Programmatic)
         return;
+    if (wasScrolledByUser)
```
Won't the indicator still flash on every stylesheet load with this change?

megangardner replied:
No, it will only flash when it's loaded the first time, not after the style sheet loads. We do still want it to be re-centered, but we don't want it to re-flash.

smfr commented on LayoutTests/http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html:
```diff
@@ -0,0 +1,137 @@
+
+<body>
+  <div id="target">
+    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vivamus lorem urna, rhoncus et erat eu, imperdiet maximus nisi. Sed rhoncus, sapien ac fringilla vehicula, ante orci vestibulum ex, eu tristique elit diam at nibh. In eget lorem massa. Nunc vel ultrices augue. Phasellus gravida neque ac maximus lobortis. Nullam vitae auctor arcu. Fusce dictum feugiat elit, at dapibus nisl tempus eget ...
```
No need for all this text. Just use `<div style="height: 500px>" or something.

smfr commented on LayoutTests/http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html:
```diff
@@ -0,0 +1,96 @@
+
+<head><script src="/js-test-resources/ui-helper.js"></script></head>
+
+<script>
```
This script tag should be inside the `<head>`.

smfr commented on LayoutTests/http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html:
```diff
@@ -0,0 +1,96 @@
+
+    if (window.testRunner)
+        testRunner.notifyDone();
+  }
```
Indentation of this brace does not match the opening brace; weird indentation in general.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -4665,6 +4665,8 @@ void LocalFrameView::setWasScrolledByUser(bool wasScrolledByUser)
     cancelScheduledScrolls();
     if (currentScrollType() == ScrollType::Programmatic)
         return;
+    if (wasScrolledByUser)
```
I'd put a blank line above line 4668 and one after 4669.

megangardner committed:
Scroll To Text Fragement re-loads after every dynamic style sheet load, and fights user scrolls to do so.
https://bugs.webkit.org/show_bug.cgi?id=264195
rdar://112608578

Reviewed by Simon Fraser.

In order to account for style changes adjusting the scroll of the page, after each dynamic style
sheet load, we recall scrollToFragment to set the page at the correct position again.

We need to stop automatically scrolling after the user has scrolled, as user scroll should trump
all URL programatic scrolls. So we reset the flag that causes scrolls to happen when a user scroll is
detected.

* LayoutTests/http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load-expected.txt: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async initiateUserScroll):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didRemoveAllPendingStylesheet):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragment):
(WebCore::LocalFrameView::setWasScrolledByUser):
* Source/WebCore/page/LocalFrameView.h:

Canonical link: https://commits.webkit.org/272151@main
